### ISSUE TITLE
C0103/N815: Constant names are not checked against UPPER_CASE naming convention

### ISSUE THREAD

**Issue by Fall1ngStar** [charliermarsh/ruff#2964]
Hi ruff team,

It looks like ruff doesn't check the name of constants like pylint does, despite `invalid-name` / `C0103` (`N815`) being marked as done in the [pylint tracking issue](https://github.com/charliermarsh/ruff/issues/970).

PEP 8 recommends UPPER_CASE for constant names ([ref](https://peps.python.org/pep-0008/#constants)).

### Example

```py
# file.py
from typing import NamedTuple

Other = NamedTuple("Other", "a,b,c")

not_correct = "tata"
CORRECT = ""
ALSO_CORRECT = ""

def foo():
    print(not_correct)
```
```sh
$ pylint file.py --disable=all --enable=C0103
************* Module file
file.py:6:0: C0103: Constant name "not_correct" doesn't conform to UPPER_CASE naming style (invalid-name)

------------------------------------------------------------------
Your code has been rated at 8.57/10 (previous run: 8.57/10, +0.00)

$ ruff file.py --verbose --no-cache
[2023-02-16][12:57:12][ruff::commands::run][DEBUG] Identified files to lint in: 226.791µs
[2023-02-16][12:57:12][ruff::commands::run][DEBUG] Checked 1 files in: 7.339083ms

$ ruff --version
ruff 0.0.247
```

### Solutions

pep8-naming N816 (mixedCase variable in global scope) already implements a weaker version of this, so I was able to adapt the code to only allow UPPER_CASE variable names in global scope, but the two rules feel a bit redundant together as some constant named `thisIs_incorrect` would trigger both N816 (because it is a mixed case naming) and C0103 (it does not follow UPPER_CASE convention for constants).

I guess there are 2 paths forward to solve this issue:
- Implement C0103 as a separate rule, but accept that some variable naming may trigger 2 separate diagnostics
- Modify N816 to only allow UPPER_CASE names, but this would differ from the original `pep8-naming` package

In any case, I would be glad to submit a PR to help get this fixed !

_PS: Thanks for the awesome work you're all doing !_

**Comment by charliermarsh** [comment]
I'm not sure how I want to proceed, but I found this old issue on `pep8-naming`: https://github.com/PyCQA/pep8-naming/issues/49.

I suppose I'd be okay with implementing C0103 as a separate rule.

**Comment by Fall1ngStar** [comment]
I will work on that if that's ok !

**Comment by Fall1ngStar** [comment]
👋 Looking for a bit of guidance here, should I implement the full [`invalid-name`](https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/invalid-name.html) from pylint ? With the checks for every type of names ? There seems to be a lot of overlap with the rules from pep8-naming so I just want to make sure that the scope for this issue is correct before I work on it 😄

**Comment by woodconn** [comment]
While there are already some N801, N802, etc. rules, it would indeed be ideal to the full functionality of this as it's own pylint category of rules.  That way people can mix and match as they see fit.

**Comment by RmStorm** [comment]
What's the current status on this? I would also really like a rule to enforce upper case constants in the global scope! I think the problem @Fall1ngStar ran into is that [C0103](https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/invalid-name.html) has a lot of overlap with the [pep8-naming (N)](https://docs.astral.sh/ruff/rules/#pep8-naming-n) rules..

Actually implementing C0103 seems to come down to implementing this [list of rules from pylint](https://github.com/pylint-dev/pylint/blob/512c8bee6cd237a8e56ab5527ee2876d8070c0d6/pylint/checkers/base/name_checker/naming_style.py#L121C1-L133C2), where every single name in python has a precisely acceptable format. Seems close to incompatible with the (N) rules but maybe they actually are compatible?

Would a PR implementing C0103 in full be welcome even though it's a rule with a rather large scope?
Or is a smaller rule only implementing upper case checking on global constant be more desirable?

**Comment by ansiblejunky** [comment]
+1 I'd also like to see this rule implemented in order to enforce the pylint C0103 rule.

**Comment by jack-mcivor** [comment]
> Or is a smaller rule only implementing upper case checking on global constant be more desirable?

I think a partial implementation is still really helpful

From the [C0103 docs](https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/invalid-name.html) and [styles](https://github.com/pylint-dev/pylint/blob/512c8bee6cd237a8e56ab5527ee2876d8070c0d6/pylint/checkers/base/name_checker/naming_style.py#L121C1-L133C2), there are two cases that require UPPER_CASE:
1. const: Module-level constants: any name defined at module level that is not bound to a class object nor reassigned.
2. class-const: Enum constants and class variables annotated with Final

Is it easier and less contentious to implement the check just on class-const names? Or even just on enum constants?
